Add locking functionality to redstone repeater

ollyhayes opened the pull request:
I've finished implementing this and it seems to work. I know you're busy so I've deployed a test version [here](https://minecraft.olly.fr.to) for quick testing (it's just hosted as a static page as I'm not too familiar with PHP, so none of the saving/loading etc will work I think).

![repeater](https://user-images.githubusercontent.com/6783708/40587959-1f869f5a-619c-11e8-9a59-9f07ecbfcdd2.png)

I put the locked/unlocked boolean state in the repeater block's metadata, this isn't the same as how it works in actual Minecraft, where it's stored in the [block state rather than block data](https://minecraft.gamepedia.com/Redstone_Repeater#Block_data), but I don't think there is such a thing as block state in the simulator right? The metadata now has the rightmost bits as direction (00011), the next two bits are delay (01100), and the leftmost bit is the new locked state (10000).

In order to get this working I had to increase the max block metadata size limit [in JsNbtParser/World_Schematic](https://github.com/JonathanLydall/JsNbtParser/blob/ae89c21b30bade872eb3f3eb4e087cb94ebbc39f/js/World_Schematic.js#L193c), because the limit is currently set to 0xf. Though the error message does say value must be from 0 to 127 (0x1f) so maybe this is okay? In my [test branch](https://github.com/ollyhayes/JavaScript-Redstone-Simulator/branches) I've pulled JsNbtParser into the repository to get it working but for the live version I'm not quite sure how it works because the script tags in [/src/_fileList.txt](https://github.com/ollyhayes/JavaScript-Redstone-Simulator/blob/d950bca4892fe961b7365284c01b7a7c25cbec11/src/_fileList.txt#L10) don't seem to match the files in your [JsNbtParser repository](https://github.com/JonathanLydall/JsNbtParser) on github. Should I submit a pull request to JsNbtParser with that change too?.

Let me know what you think, happy to make any changes you think are necessary.

(This fixes issue #1)

ollyhayes committed:
Add locked bit to repeater metadata

ollyhayes committed:
Don't schedule update if repeater is locked

ollyhayes committed:
Add grey bar to show when repeater is locked

ollyhayes committed:
Only lock if locking repeater is pointing in the right direction

frogg commented:
Great to see that someone is still working on this 😃👍

JonathanLydall commented:
Impressive amount of work, especially as my code was not the best due to
relative inexperience at the time.

I’m worried about the metadata size change as it will break compatibility
with reading/saving to schematic files for MCEdit (unless the format has
changed).

According to the wiki, at least initially, block state is not persisted to
save files, so perhaps we don’t have to worry about that. However, due to
the flyweight pattern, we need to be able to determine somehow whether or
not it is locked.

I suspect the easiest way to do this is to read the state of the blocks to
either side of it. An isLocked method would be an if statement checking
whether the orientation is north/south else east/west, then reading the
block on either side, to see if they are repeaters in the correct state.

For all the current blocks in the simulator they are actually using the
exact same logic as the version of Minecraft out at that time, as I
referred to the decompiled source code. Unfortunately the engine has
changed drastically since then, so source code may not be possible to
integrate with the simulator as it stands. However, I think this behaviour
is simple enough to get perfect without having to refer to the game’s code.

I moved recently and am still waiting for internet to be installed at home,
but I hope to be online this week, otherwise no later than sometime the
week after.

Thanks for the efforts, I will absolutely get any changes live once they
are implemented in a way that doesn’t break anything.


On Sun, 27 May 2018 at 17:57, ollyhayes <notifications@github.com> wrote:

> I've finished implementing this and it seems to work. I know you're busy
> so I've deployed a test version here <https://minecraft.olly.fr.to> for
> quick testing (it's just hosted as a static page as I'm not too familiar
> with PHP, so none of the saving/loading etc will work I think).
>
> [image: repeater]
> <https://user-images.githubusercontent.com/6783708/40587959-1f869f5a-619c-11e8-9a59-9f07ecbfcdd2.png>
>
> I put the locked/unlocked boolean state in the repeater block's metadata,
> this isn't the same as how it works in actual Minecraft, where it's stored
> in the block state rather than block data
> <https://minecraft.gamepedia.com/Redstone_Repeater#Block_data>, but I
> don't think there is such a thing as block state in the simulator right?
> The metadata now has the rightmost bits as direction (00011), the next two
> bits are delay (01100), and the leftmost bit is the new locked state
> (10000).
>
> In order to get this working I had to increase the max block metadata size
> limit in JsNbtParser/World_Schematic
> <https://github.com/JonathanLydall/JsNbtParser/blob/ae89c21b30bade872eb3f3eb4e087cb94ebbc39f/js/World_Schematic.js#L193c>,
> because the limit is currently set to 0xf. Though the error message does
> say value must be from 0 to 127 (0x1f) so maybe this is okay? In my test
> branch
> <https://github.com/ollyhayes/JavaScript-Redstone-Simulator/branches>
> I've pulled JsNbtParser into the repository to get it working but for the
> live version I'm not quite sure how it works because the script tags in
> /src/_fileList.txt
> <https://github.com/ollyhayes/JavaScript-Redstone-Simulator/blob/d950bca4892fe961b7365284c01b7a7c25cbec11/src/_fileList.txt#L10>
> don't seem to match the files in your JsNbtParser repository on github.
> Should I submit a pull request to JsNbtParser with that change too?.
>
> Let me know what you think, happy to make any changes you think are
> necessary.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/JonathanLydall/JavaScript-Redstone-Simulator/pull/20
> Commit Summary
>
>    - Add locked bit to repeater metadata
>    - Don't schedule update if repeater is locked
>    - Add grey bar to show when repeater is locked
>    - Only lock if locking repeater is pointing in the right direction
>
> File Changes
>
>    - *M* src/blocks/RedstoneRepeater.js
>    <https://github.com/JonathanLydall/JavaScript-Redstone-Simulator/pull/20/files#diff-0>
>    (46)
>
> Patch Links:
>
>    -
>    https://github.com/JonathanLydall/JavaScript-Redstone-Simulator/pull/20.patch
>    -
>    https://github.com/JonathanLydall/JavaScript-Redstone-Simulator/pull/20.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JonathanLydall/JavaScript-Redstone-Simulator/pull/20>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOwFm10-uYxnI-FEcbsbLPDX4sF57KXks5t2szbgaJpZM4UPVHT>
> .
>

ollyhayes commented:
> we need to be able to determine somehow whether or
not it is locked

Perhaps I'm misunderstanding, but I think the isLocked method you're describing is the one I created [here](https://github.com/ollyhayes/JavaScript-Redstone-Simulator/blob/d950bca4892fe961b7365284c01b7a7c25cbec11/src/blocks/RedstoneRepeater.js#L123) in RedstoneRepeater.js:

```javascript
	proto.isLocked = function (world, posX, posY, posZ, blockMetadata) {
		var repeaterDirection= blockMetadata & 3;
		var repeaterBlockId = world.Block.redstoneRepeaterActive.blockID;
		switch (repeaterDirection) {
			case 3:
			case 1:
				return (world.getBlockId(posX, posY, posZ + 1) === repeaterBlockId && world.isBlockProvidingPowerTo(posX, posY, posZ + 1, 3))
					|| (world.getBlockId(posX, posY, posZ - 1) === repeaterBlockId && world.isBlockProvidingPowerTo(posX, posY, posZ - 1, 2));

			case 0:
			case 2:
				return (world.getBlockId(posX + 1, posY, posZ) === repeaterBlockId && world.isBlockProvidingPowerTo(posX + 1, posY, posZ, 5))
					|| (world.getBlockId(posX - 1, posY, posZ) === repeaterBlockId && world.isBlockProvidingPowerTo(posX - 1, posY, posZ, 4));
		}
		return false;
	}
```

This is called when any neighbor of the repeater is changed and the metadata is updated based on the outcome of this function.

No rush anyway, happy to wait till you get internet sorted for a response.

JonathanLydall commented:
Had a chance to actually look at the diffs from a computer now, before my comments were done on my phone and based on your comments only.

Your isLocked method is exactly what I was thinking we would need. We do however want to get rid of all the metadata changes for the reasons I discussed before. If recall correctly from when I implemented everything, we shoudn't need to update and notify on lock/unlock because the rendering logic will be called anyway if a neighouring block is changed.

ollyhayes committed:
Revert "Add locked bit to repeater metadata"

This reverts commit 41cacca124cde0d5cdd28a896943c224af5acc79.

ollyhayes committed:
Don's use metadata for locked state, just check directly

ollyhayes commented:
Ah yeah that makes sense, I never actually needed to store the locked state. I've removed the metadata stuff, and now in the draw method I'm just calling isLocked rather than checking the data.

Updated the [test deployment](https://minecraft.olly.fr.to) and it still works :)

ollyhayes commented:
Had any chance to look at merging this yet? I've been using it myself for a while now and not had any issues.

Here's a counter I created using repeater locking. Being able to quickly prototype this is the reason I wanted to be able to lock repeaters:

<img src="https://user-images.githubusercontent.com/6783708/41233401-f41bbada-6d4e-11e8-8f3f-04a0746efe27.png" width="200">

My test version doesn't have any ability to save it though, so looking forward to the real one being updated!

JonathanLydall commented:
Just wanted to say I haven't forgotten about this, I hit challenges on my development environment and my time has been tight for a while (spare time project, full time job and upcoming wedding). I'm hoping to make progress soon.

JonathanLydall commented:
I have uploaded this into production, your changes can now be saved.

JonathanLydall commented:
Closes #1.
